sql/src/main/java/io/crate/executor/transport/task/UpdateByIdTask.java

```diff
@@ -0,0 +1,155 @@
+                          UpdateById updateById) {
+        super(updateById.jobId());
+        this.clusterService = clusterService;
+        this.functions = functions;
+        this.shardUpsertAction = shardUpsertAction;
+        this.updateById = updateById;
+        assignments = Assignments.convert(updateById.assignmentByTargetCol());
+        createBuilder = (continueOnError) ->
+            new ShardUpsertRequest.Builder(
+                ShardingUpsertExecutor.BULK_REQUEST_TIMEOUT_SETTING.setting().get(clusterService.state().metaData().settings()),
+                false,
+                continueOnError,
+                assignments.targetNames(),
+                null, // missing assignments are for INSERT .. ON DUPLICATE KEY UPDATE
+                updateById.jobId(),
+                false
+            );
+    }
+
+    @Override
+    public void execute(RowConsumer consumer, Row parameters) {
+        HashMap<ShardId, ShardUpsertRequest> requestsByShard = new HashMap<>();
+        ShardUpsertRequest.Builder requestBuilder = createBuilder.apply(updateById.docKeys().size() > 1);
+        addRequests(0, parameters, requestsByShard, requestBuilder);
+        MultiActionListener<ShardResponse, long[], ? super Row> listener = new MultiActionListener<>(
+            requestsByShard.size(),
+            () -> new long[]{0},
+            DeleteByIdTask::updateRowCountOrFail,
+            rowCount -> new Row1(rowCount[0]),
+            new OneRowActionListener<>(consumer, Function.identity())
+        );
+        for (ShardUpsertRequest request : requestsByShard.values()) {
+            shardUpsertAction.execute(request, listener);
+        }
+    }
+
+    @Override
+    public List<CompletableFuture<Long>> executeBulk(List<Row> bulkParams) {
+        HashMap<ShardId, ShardUpsertRequest> requestsByShard = new HashMap<>();
+        ShardUpsertRequest.Builder requestBuilder = createBuilder.apply(true);
+        int location = 0;
+        // idx=location; value=resultIdx
+        IntArrayList bulkIndices = new IntArrayList(bulkParams.size() * updateById.docKeys().size());
+        for (int resultIdx = 0; resultIdx < bulkParams.size(); resultIdx++) {
+            int prevLoc = location;
+            location = addRequests(location, bulkParams.get(resultIdx), requestsByShard, requestBuilder);
+            for (int i = prevLoc; i < location; i++) {
+                bulkIndices.add(resultIdx);
+            }
+        }
+        BulkShardResponseListener listener =
+            new BulkShardResponseListener(requestsByShard.size(), bulkParams.size(), bulkIndices);
+        for (ShardUpsertRequest request : requestsByShard.values()) {
+            shardUpsertAction.execute(request, listener);
+        }
+        return listener.rowCountFutures();
+    }
+
+    private int addRequests(int location, Row parameters, HashMap<ShardId, ShardUpsertRequest> requestsByShard, ShardUpsertRequest.Builder requestBuilder) {
```

Comment:
could we maybe share this logic with DeleteByIdTask as well? also is it require to expect a HashMap here?